### PR TITLE
t-ronmoneta/cs sample app refactored

### PR DESCRIPTION
**Pull request by ronmonetaMicro**
Post skeleton talk refactoring of C# sampleApp

### CODE REVIEW

File: client/DatabaseCursorSample/DatabaseCursorSample.csproj [hunk 1]
```diff
@@ -1,132 +1,14 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (AsafMah) [inline client/DatabaseCursorSample/DatabaseCursorSample.csproj:1]: What happened here? why did you change all of the projects?

File: client/QuickStart/API_Util.cs [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        public int WaitForIngestSeconds { get; set; }
+
+        /// Optional - Customized ingestion batching policy
+        public string BatchingPolicy { get; set; }
+    }
+
+    /// <summary>
+    /// Util static class - Handels the communication with the API, and provides generic and simple "plug-n-play" functions to use in different programs.
+    /// </summary>
+    public static class Util
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: We usually name files after classes without underscores, so you can just call this "Util.cs", though we usually use the plural Utils

Also, I'm not a big fan of this class - there isn't a unifying theme, it does all sorts of stuff. Maybe split it to different kinds of utils

File: client/QuickStart/kusto_sample_config.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{
+	"kustoUri" : "https://kvc8145132f55d24729976.northeurope.kusto.windows.net",
```
Comment (AsafMah) [inline client/QuickStart/kusto_sample_config.json]: Don't commit the file with your own cluster. In java we use `https://sdkse2etest.eastus.kusto.windows.net` which isn't great either, we probably want `https://help.kusto.windows.net`

File: client/QuickStart/SampleApp.cs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    /// adapting the code to your needs.
+    /// </summary>
+    public static class KustoSampleApp
+    {
+
+        // TODO (config):
+        // If this quickstart app was downloaded from OneClick, kusto_sample_config.json should be pre-populated with your cluster's details.
+        // If this quickstart app was downloaded from GitHub, edit kusto_sample_config.json and modify the cluster URL and database fields appropriately.
+        private const string ConfigFileName = @"kusto_sample_config.json";
+        private static int _step = 1;
```
Comment (AsafMah) [inline client/QuickStart/SampleApp.cs]: Naming - why is step not capital but WaitForUser is? why start with an underscore at all?

File: client/QuickStart/SampleApp.cs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+        /// <summary>
+        /// Second phase - The ingestion process.
+        /// </summary>
+        /// <param name="config">ConfigJson object</param>
+        /// <param name="adminClient">Privileged client to run Control Commands</param>
+        /// <param name="ingestClient">Client to ingest data</param>
+        private static async Task IngestionAsync(ConfigJson config, ICslAdminProvider adminClient, IKustoIngestClient ingestClient)
+        {
+            foreach (var file in config.Data)
```
Comment (AsafMah) [inline client/QuickStart/SampleApp.cs]: Make this into a proper class too, instead of magic strings and parsing?

File: client/QuickStart/SampleApp.cs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        /// </summary>
+        /// <param name="useExistingMapping">Flag noting if we should the existing mapping or create a new one</param>
+        /// <param name="adminClient">Privileged client to run Control Commands</param>
+        /// <param name="configDatabaseName">DB name</param>
+        /// <param name="configTableName">Table name</param>
+        /// <param name="mappingName">Desired mapping name</param>
+        /// <param name="mappingValue">Values of the new mappings to create</param>
+        /// <param name="dataFormat">Given data format</param>
+        /// <returns>True if Ingestion Mappings exists (whether by us, or the already existing one)</returns>
+        private static async Task<bool> CreateIngestionMappingsAsync(bool useExistingMapping, ICslAdminProvider adminClient, string configDatabaseName, string configTableName, string mappingName, string mappingValue, DataSourceFormat dataFormat)
```
Comment (AsafMah) [inline client/QuickStart/SampleApp.cs]: Why do you completely swallow the exception here? even if it's valid to continue after, at least show what it was

File: client/QuickStart/SampleApp.cs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+            WaitForUserToProceed($"Ingest '{sourceUri}' from '{sourceType}'");
+
+            // Tip: When ingesting json files, if each line represents a single-line json, use MULTIJSON format even if the file only contains one line.
+            // If the json contains whitespace formatting, use SINGLEJSON. In this case, only one data row json object is allowed per file.
+            dataFormat = dataFormat == DataSourceFormat.json ? DataSourceFormat.multijson : dataFormat;
+
+            // Tip: Kusto's C# SDK can ingest data from files, blobs and open streams.See the SDK's samples and the E2E tests in azure.kusto.ingest for additional references.
+            switch (sourceType)
+            {
+                case "localfilesource":
```
Comment (AsafMah) [inline client/QuickStart/SampleApp.cs]: Maybe make this an enum too?

File: client/QuickStart/API_Util.cs [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+                {
+                    (name: nameof(config.DatabaseName), value: config.DatabaseName),
+                    (name: nameof(config.TableName), value: config.TableName),
+                    (name: nameof(config.TableSchema), value: config.TableSchema),
+                    (name: nameof(config.KustoUri), value: config.KustoUri),
+                    (name: nameof(config.IngestUri), value: config.IngestUri),
+                    (name: nameof(config.AuthenticationMode), value: config.AuthenticationMode)
+                }.Where(item => string.IsNullOrWhiteSpace(item.value)).ToArray();
+
+                if (missing.Any())
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: Style - always brackets after if

File: client/QuickStart/API_Util.cs [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+                if (config.Data is null || !config.Data.Any() || config.Data[0].Count == 0)
+                    ErrorHandler($"Required field Data in '{configFilePath}' is either missing, empty or misfilled");
+                return config;
+            }
+            catch (Exception ex)
+            {
+                ErrorHandler($"Couldn't read config file: '{configFilePath}'", ex);
+            }
+
+            return null;
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: Thinking about it, I don't like returning null in this state. If we had a bug and it reached here, we will get a subtle NRE later.
Maybe instead do `throw new InvalidOperationException("Unreachable code");`?

File: client/QuickStart/kusto_sample_config.json [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	"ingestUri" : "https://ingest-kvc8145132f55d24729976.northeurope.kusto.windows.net",
+	"databaseName" : "MyDatabase",
+	"tableName" : "SampleTable",
+	"useExistingTable": true,
+	"alterTable": true,
+	"queryData": true,
+	"ingestData": true,
+	"AuthenticationMode" :"UserPrompt",
```
Comment (AsafMah) [inline client/QuickStart/kusto_sample_config.json]: Formatting, the colons are inconsistent

Comment (AsafMah) [inline client/QuickStart/kusto_sample_config.json]: Enum?

File: client/QuickStart/API_Util.cs [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+            {
+                (name: nameof(certificatePath), value: certificatePath),
+                (name: nameof(certificatePassword), value: certificatePassword),
+                (name: nameof(applicationId), value: tenantId),
+            }.Where(item => string.IsNullOrWhiteSpace(item.value)).ToArray();
+            if (missing.Any())
+                ErrorHandler($"Missing the following required fields in configuration file in order to authenticate using a certificate: {string.Join(", ", missing.Select(item => item.name))}");
+
+            var appId = Environment.GetEnvironmentVariable("APP_ID");
+            var SubjectDistinguishedName = Environment.GetEnvironmentVariable("SUBJECT_DISTINGUISHED_NAME");
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: naming - don't start local with capital

File: client/QuickStart/API_Util.cs [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+        /// </summary>
+        /// <param name="clusterUrl">Url of cluster to connect to</param>
+        /// <param name="certificatePath">Given certificate path</param>
+        /// <param name="certificatePassword">Given certificate password</param>
+        /// <param name="applicationId">Given application id</param>
+        /// <param name="tenantId">Given tenant id</param>
+        /// <returns>AppCertificate Kusto Connection String</returns>
+        public static KustoConnectionStringBuilder CreateAppCertificateConnectionString(string clusterUrl, string certificatePath, string certificatePassword, string applicationId, string tenantId)
+        {
+            var missing = new[]
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: I thought we are only doing issuer - you don't need these

File: client/QuickStart/API_Util.cs [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+            catch (InvalidOperationException e)
+            {
+                ErrorHandler($"Failed to load PEM file from {privateKeyPemFilePath}", e);
+            }
+
+            if (publicCertFilePath != null)
+            {
+                try
+                {
+                    publicCertificate = File.ReadAllText(publicCertFilePath);
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: You don't use this and don't need this

File: client/QuickStart/API_Util.cs [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+            var SubjectDistinguishedName = Environment.GetEnvironmentVariable("SUBJECT_DISTINGUISHED_NAME");
+            var IssuerDistinguishedName = Environment.GetEnvironmentVariable("ISSUER_DISTINGUISHED_NAME");
+            var privateKeyPemFilePath = Environment.GetEnvironmentVariable("PRIVATE_KEY_PEM_FILE_PATH");
+            var publicCertFilePath = Environment.GetEnvironmentVariable("PUBLIC_CERT_FILE_PATH");
+            string publicCertificate;
+            string pemCertificate;
+
+            try
+            {
+                pemCertificate = File.ReadAllText(privateKeyPemFilePath ?? throw new InvalidOperationException());
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: You don't use this and don't need this

File: client/QuickStart/API_Util.cs [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+                }
+                catch (InvalidOperationException e)
+                {
+                    ErrorHandler($"Failed to load public certificate file from {publicCertFilePath}", e);
+                }
+                return new KustoConnectionStringBuilder(clusterUrl).WithAadApplicationSubjectAndIssuerAuthentication(appId, SubjectDistinguishedName, IssuerDistinguishedName, tenantId);
+            }
+
+
+            X509Certificate2 certificate = null;
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: Remove - we don't want this here (right @yochai?)

File: client/QuickStart/API_Util.cs [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+        /// </summary>
+        /// <param name="client">Premade client to run Commands. can be either an adminClient or queryClient</param>
+        /// <param name="configDatabaseName">DB name</param>
+        /// <param name="command">The Command to execute</param>
+        /// <returns>True on success, false otherwise</returns>
+        public static async Task ExecuteAsync(IDisposable client, string configDatabaseName, string command)
+        {
+            try
+            {
+                Newtonsoft.Json.Linq.JObject[] result;
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: Don't need fully qualified name

File: client/QuickStart/API_Util.cs [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+                    var clientRequestProperties = CreateClientRequestProperties("CS_SampleApp_Query");
+                    ICslQueryProvider queryClient = (ICslQueryProvider)client;
+                    result = (await queryClient.ExecuteQueryAsync(configDatabaseName, command, clientRequestProperties)).ToJObjects().ToArray();
+                }
+
+                // Tip: Actual implementations wouldn't generally print the response from a control command or a query .We print here to demonstrate what a sample of the response looks like.
+                // Moreover, there are some built-in classes for control commands under the Kusto.Data namespace -for example, Kusto.Data.TablesShowCommandResult maps to the result of the ".show tables" commands
+                Console.WriteLine($"Response from executed command '{command}':\n--------------------");
+                var firstRow = result[0];
+                foreach (var item in firstRow.Properties())
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: brackets

File: client/QuickStart/API_Util.cs [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+                    Console.WriteLine(item);
+            }
+            catch (Exception ex)
+            {
+                var err = ex.GetType().ToString();
```
Comment (AsafMah) [inline client/QuickStart/API_Util.cs]: DO NOT DO THIS - this isn't python, don't compare types by name.
use `if ex is KustoClientException kcs { ...}`

(same for other places)

**Review by AsafMah — CHANGES_REQUESTED** [review]
Some notes

File: client/DatabaseCursorSample/DatabaseCursorSample.csproj (hunk 1 above)
Comment (ronmonetaMicro) [reply client/DatabaseCursorSample/DatabaseCursorSample.csproj:1]: Compatibility problems that were discovered when using VSCode. the projects had different version of using packages - that VSStudio could handle by itself - but Code didnt know how to work with the older version. Discussed it with @yogilad and changed it.

File: client/QuickStart/API_Util.cs (hunk 2 above)
Comment (ronmonetaMicro) [reply client/QuickStart/API_Util.cs]: Name change - done.
This is what we discussed previously - create one class to do the proper SampleApp actions, and one that handles all the API calls and envelops it for the user. i can try and split it further (maybe into two utils - one for execution and one for ingestion), but ill admit i think it will be a bit redundant, as it will start to be a lot like the SDK itself.

File: client/QuickStart/SampleApp.cs (hunk 5 above)
Comment (ronmonetaMicro) [reply client/QuickStart/SampleApp.cs]: not quite sure what you meant.  a class of what? dataFormat?

File: client/QuickStart/SampleApp.cs (hunk 6 above)
Comment (ronmonetaMicro) [reply client/QuickStart/SampleApp.cs]: Utils.ExecuteAsync which is called there handles the exception internally, that catch block is there only due to syntactical requirements. perhaps i should just remove the try catch block here. WDYT?

File: client/QuickStart/SampleApp.cs (hunk 7 above)
Comment (ronmonetaMicro) [reply client/QuickStart/SampleApp.cs]: Ok. Do we however want to provide more options then? do we want the user to be able to expand upon it?

File: client/QuickStart/kusto_sample_config.json (hunk 10 above)
Comment (ronmonetaMicro) [reply client/QuickStart/kusto_sample_config.json]: In the config json file? wdym?

File: client/QuickStart/API_Util.cs (hunk 12 above)
Comment (ronmonetaMicro) [reply client/QuickStart/API_Util.cs]: You do have thought the option to use the X509 certificate if publicCertFilePath is null. do we wish to remove it and stay only with issuer?

File: client/QuickStart/API_Util.cs (hunk 15 above)
Comment (ronmonetaMicro) [reply client/QuickStart/API_Util.cs]: https://github.com/Azure/azure-kusto-samples-dotnet/pull/19#discussion_r884427197
converging with this discussion

File: client/QuickStart/API_Util.cs (hunk 16 above)
Comment (ronmonetaMicro) [reply client/QuickStart/API_Util.cs]: does not recognize it without. do i need to adjust something else somewhere first?

File: client/QuickStart/API_Util.cs [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+using Newtonsoft.Json;
+using Kusto.Cloud.Platform.Data;
+using ShellProgressBar;
+
+namespace QuickStart
+{
+    /// <summary>
+    /// ConfigJson object - represents a cluster and DataBase connection configuration file.
+    /// </summary>
+    public class ConfigJson
```
Comment (yogilad) [inline client/QuickStart/API_Util.cs]: This class and it's read method should be part of the app itself (not a reusable kusto code)

File: client/QuickStart/API_Util.cs (hunk 15 above)
Comment (yochai) [reply client/QuickStart/API_Util.cs]: > Remove - we don't want this here (right @yochai?)

@AsafMah, you pinged the wrong user

File: client/QuickStart/API_Util.cs (hunk 2 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: I'll talk to Yochai too and we'll see.

File: client/QuickStart/SampleApp.cs (hunk 4 above)
Comment (AsafMah) [reply client/QuickStart/SampleApp.cs]: Now - why _WaitForUser still has the underscore?

File: client/QuickStart/SampleApp.cs (hunk 5 above)
Comment (AsafMah) [reply client/QuickStart/SampleApp.cs]: All of the properties inside config.Data

```c#
        [JsonObject(NamingStrategyType = typeof(CamelCaseNamingStrategy))]

        class FileData
        {
            private string Format { get; set; };
            ...
        }
```

File: client/QuickStart/SampleApp.cs (hunk 6 above)
Comment (AsafMah) [reply client/QuickStart/SampleApp.cs]: Yes

File: client/QuickStart/SampleApp.cs (hunk 7 above)
Comment (AsafMah) [reply client/QuickStart/SampleApp.cs]: We don't want to provide anything else right now.

File: client/QuickStart/kusto_sample_config.json (hunk 10 above)
Comment (AsafMah) [reply client/QuickStart/kusto_sample_config.json]: No, in the c# define an enum for the Authentication Mode values, and use it in the config class instead of a string. You get free validation too.

File: client/QuickStart/API_Util.cs (hunk 12 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: Yes

File: client/QuickStart/API_Util.cs (hunk 15 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: @yogilad

File: client/QuickStart/API_Util.cs (hunk 16 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: You need a using. Try to see if the IDE/vscode can do it for you.

File: client/QuickStart/API_Util.cs (hunk 18 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: Good. but brackets here and elsewhere in the file

File: client/QuickStart/API_Util.cs (hunk 19 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: Maybe move this to a third file?

File: client/QuickStart/SampleApp.cs (hunk 5 above)
Comment (ronmonetaMicro) [reply client/QuickStart/SampleApp.cs]: Do you mean in a similar way to what we did in TS?
image

what added value will we get here? we would still have to do the parsing somewhere

Comment (AsafMah) [reply client/QuickStart/SampleApp.cs]: If you'll embed it in ConfigJson (`public List<ConfigData> Data { get; set; }`), it will parse it automatically when you read the config.

File: client/DatabaseCursorSample/DatabaseCursorSample.csproj (hunk 1 above)
Comment (yogilad) [reply client/DatabaseCursorSample/DatabaseCursorSample.csproj:1]: yes. though in retrospect this was worthy of it's own PR.
My bad, not Ron's

File: client/QuickStart/Utils.cs [hunk 20]
```diff
@@ -194,43 +95,27 @@ public static KustoConnectionStringBuilder CreateAppCertificateConnectionString(
                 ErrorHandler($"Missing the following required fields in configuration file in order to authenticate using a certificate: {string.Join(", ", missing.Select(item => item.name))}");
 
             var appId = Environment.GetEnvironmentVariable("APP_ID");
-            var SubjectDistinguishedName = Environment.GetEnvironmentVariable("SUBJECT_DISTINGUISHED_NAME");
-            var IssuerDistinguishedName = Environment.GetEnvironmentVariable("ISSUER_DISTINGUISHED_NAME");
-            var privateKeyPemFilePath = Environment.GetEnvironmentVariable("PRIVATE_KEY_PEM_FILE_PATH");
+            var subjectDistinguishedName = Environment.GetEnvironmentVariable("SUBJECT_DISTINGUISHED_NAME");
```
Comment (yogilad) [inline client/QuickStart/Utils.cs]: Since this is batched together, it's worth saying which variables affect which flow, e.g. by introducing another condition for subject name even though programmatically it's not needed

File: client/QuickStart/Utils.cs [hunk 21]
```diff
@@ -285,15 +170,16 @@ public static async Task ExecuteAsync(IDisposable client, string configDatabaseN
...
+                {
                     Console.WriteLine(item);
+                }
             }
             catch (Exception ex)
             {
-                var err = ex.GetType().ToString();
                 string msg;
-                if (err == "KustoClientException")
+                if (ex is KustoClientException)
```
Comment (yogilad) [inline client/QuickStart/Utils.cs]: The better practice for this is to have a set of catch blocks with explicit types.
@AsafMah , do you agree?

File: client/QuickStart/API_Util.cs (hunk 16 above)
Comment (ronmonetaMicro) [reply client/QuickStart/API_Util.cs]: VSCODE is incapable of such stuff.
anyway, seems like JObject is not IDisposable - so we cant really use a using block here (see here: https://stackoverflow.com/questions/42077388/type-used-in-a-using-statement-must-be-implicitly-convertible-to-system-idispos) 
 
image

File: client/QuickStart/Utils.cs (hunk 20 above)
Comment (ronmonetaMicro) [reply client/QuickStart/Utils.cs]: Not needed as we removed the x509 option as mentioned here: https://github.com/Azure/azure-kusto-samples-dotnet/pull/19#discussion_r884427197

File: client/QuickStart/Utils.cs (hunk 21 above)
Comment (ronmonetaMicro) [reply client/QuickStart/Utils.cs]: im actually not sure why i haven't done it like that, its the common way . maybe i saw it in a sample and took it. ill fix it

**Review by yogilad — APPROVED** [review]

File: client/QuickStart/API_Util.cs (hunk 16 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: I meant `using Newtonsoft.Json.Linq;`, not using statement, we're good.

File: client/QuickStart/Utils.cs [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+                    Console.WriteLine($"Response from executed command '{command}':\n--------------------");
+                    var firstRow = result[0];
+                    foreach (var item in firstRow.Properties())
+                    {
+                        Console.WriteLine(item);
+                    }
+                }
+                catch (KustoClientException ex)
+                {
+                    ErrorHandler(string.Format("Client error while trying to execute command '{0}' on database '{1}'", command, configDatabaseName), ex);
```
Comment (AsafMah) [inline client/QuickStart/Utils.cs]: Reduce duplication, make the repeating string into a variable. plus use $ strings for interpolation

File: client/QuickStart/API_Util.cs (hunk 2 above)
Comment (AsafMah) [reply client/QuickStart/API_Util.cs]: Looks good

**Review by AsafMah — APPROVED** [review]
Looks good - one fix and you can merge
